toolkit/tools/pkg/imagecustomizerlib/customizeboot.go

```diff
@@ -359,11 +423,82 @@ func updateSELinuxCommandLineHelper(grub2Config string, selinuxMode imagecustomi
 		return "", fmt.Errorf("unknown SELinux mode (%s)", selinuxMode)
 	}
 
-	// Insert/replace/delete SELinux command-line arguments.
-	grub2Config = fmt.Sprintf("%s %s %s", grub2Config[:insertIndex], newSELinuxArgs, grub2Config[insertIndex:])
+	grub2Config, err := updateKernelCommandLineArguments(grub2Config, []string{"security", "selinux", "enforcing"},
+		newSELinuxArgs)
+	if err != nil {
+		return "", err
+	}
+
+	return grub2Config, nil
+}
+
+// Finds a set command that sets the variable with the provided name and then change the value that is set.
+func replaceSetCommandValue(grub2Config string, varName string, newValue string) (string, error) {
+	grubTokens, err := grub.TokenizeGrubConfig(grub2Config)
+	if err != nil {
+		return "", err
+	}
+
+	grubLines := grub.SplitTokensIntoLines(grubTokens)
+	setLines := grub.FindCommandAll(grubLines, "set")
+
+	// Search for all the set commands that set the variable.
+	setVarLines := [][]grub.Token(nil)
+	for _, line := range setLines {
+		if len(line) < 2 {
+			return "", fmt.Errorf("grub config has a set command that has zero args")
+		}
+
+		argToken := line[1]
+		argStringBuilder := strings.Builder{}
+
+		// Get the name of the variable being set.
+	subWordsLoop:
+		for _, subword := range argToken.SubWords {
+			switch subword.Type {
+			case grub.KEYWORD_STRING, grub.STRING:
+				argStringBuilder.WriteString(subword.Value)
+
+			case grub.QUOTED_VAR_EXPANSION, grub.VAR_EXPANSION:
+				// There is a variable expansion (e.g. $a) in the arg.
+				// If the variable is only in the value portion (i.e. after the = symbol), then the set command can
+				// still be replaced.
+				break subWordsLoop
+			}
+		}
+
+		argValue := argStringBuilder.String()
+		name, _, foundEq := strings.Cut(argValue, "=")
+		if !foundEq {
+			return "", fmt.Errorf("grub config has a set command that doesn't have an equals symbol")
+		}
+
+		// Check if the name matches.
+		if name == varName {
+			setVarLines = append(setVarLines, line)
+		}
+	}
+
+	// Ensure there is only 1 set command.
```

Comment:
would it make sense to move this validation logic to load time? Once successfully loaded, we know that the input grub is valid.

Reply:
Even if we added pre-validation steps, I would probably still keep the on-the-fly checks just as a form of sanity check.

(If you stop the search early, it'll still be O(n).)